libs/ds18b20/ds18b20.cpp

```diff
@@ -1,19 +1,119 @@
+ * be considere explicitly.
+ * 9-12 bit resolution: 94, 188, 375, 750 ms conversion time
+ */
+#if DS18B20_RESOLUTION == 9
+#define DS18B20_ERROR_TOLERANCE 0.45
+#define DS18B20_CONVERSION_DURATION 94
+#elif DS18B20_RESOLUTION == 10
+#define DS18B20_ERROR_TOLERANCE 0.20
+#define DS18B20_CONVERSION_DURATION 188
+#elif DS18B20_RESOLUTION == 11
+#define DS18B20_ERROR_TOLERANCE 0.08
+#define DS18B20_CONVERSION_DURATION 375
+#elif DS18B20_RESOLUTION == 12
+#define DS18B20_ERROR_TOLERANCE 0.02
+#define DS18B20_CONVERSION_DURATION 750
+#endif
+
+
 Ds18b20::Ds18b20(uint8_t pin)
 : m_wire{pin}
 , m_sensors{&m_wire}
+{}
+
+Ds18b20::Ds18b20(uint8_t pin, uint8_t pin_pullup)
+: m_pin_pullup{pin_pullup}
+, m_wire{pin}
+, m_sensors{&m_wire}
 {
-    m_sensors.setWaitForConversion(false);
-    m_sensors.requestTemperaturesByIndex(0);
+    pinMode(m_pin_pullup, OUTPUT);
+    set_external_pullup(true);
+}
+
+void Ds18b20::begin()
+{
+    set_external_pullup(false);
+    m_sensors.begin();
+    m_sensors.getAddress(m_address, 0);
+    reset();
+    set_external_pullup(true);
+}
+
+void Ds18b20::reset()
+{
+    set_external_pullup(false);
+    if(m_sensors.isConnected(m_address)) {
+        m_sensors.setWaitForConversion(false);
+        m_sensors.setResolution(DS18B20_RESOLUTION);
+        m_sensors.requestTemperatures();
+        m_last_seen = millis();
+    } else {
+        m_disconnected = true;
+        m_wire.reset();
+        m_sensors.begin();
+        m_sensors.getAddress(m_address, 0);
+    }
+    set_external_pullup(true);
+}
+
+unsigned int Ds18b20::elapsed()
+{
+    return millis() - m_last_seen;
+}
+
+bool Ds18b20::is_connected()
+{
+    return !m_disconnected;
 }
 
 float Ds18b20::temperature()
 {
-    if (m_sensors.isConversionComplete()) {
-        m_last_temperature = m_sensors.getTempCByIndex(0);
-        m_sensors.requestTemperaturesByIndex(0);
+    if (elapsed() > 5000) {                     // timeout for sensor connection
+        m_disconnected = true;
+        if (elapsed() / 1000 % 5 == 0) {        // try reconenct every 5s (at the moment this tries for 1s every 5s, is there a better approach attempting this only once without additional flag?)
+            reset();
+        }
+    }
+    if (elapsed() > DS18B20_CONVERSION_DURATION) {
+        set_external_pullup(false);
+        float temp_T = m_sensors.getTempC(m_address);
+
+        if ( (abs(temp_T - DS18B20_DISCONNECTED + DS18B20_ERROR_TOLERANCE / 2) > DS18B20_ERROR_TOLERANCE / 2) &&
+             (abs(temp_T + DS18B20_CONVERSION_ERROR - DS18B20_ERROR_TOLERANCE / 2) > DS18B20_ERROR_TOLERANCE / 2)) {
+            m_last_temperature = temp_T;
+            m_last_seen = millis();
+            m_sensors.requestTemperaturesByAddress(m_address);
+            m_disconnected = false;
+        }
+        set_external_pullup(true);
     }
 
     return m_last_temperature;
 }
+
+void Ds18b20::set_external_pullup(bool state)
+{
+    if (m_pin_pullup == 255) return;
```

Comment:
```suggestion
    if (m_pin_pullup == 255) {
        return;
    }
```

Reply:
Fixed in 03d51eeef524e32c61277c2647ff848bb0dc321d

Reply:
Fixed in 03d51eeef524e32c61277c2647ff848bb0dc321d